src/blk/BlockDevice.h

```diff
@@ -288,6 +288,7 @@ class BlockDevice {
   virtual int flush() = 0;
   virtual int discard(uint64_t offset, uint64_t len) { return 0; }
```

Comment:
No need to expose in BlockDevice interface now. Hence no need for declaring them virtual too.
The same for queue_discard()
And maybe prefix these names with '_' in KernelDevice since they are private now..

Reply:
Thank you very much for this detailed review.
Since BlueStore will discard the whole device at creating, I add a parameter to try_discard so caller can still choose to discard synchronously even if async discard is enabled.
Other places have been updated according to the comments.